File: leetcode/118.cpp
// Runtime: 3 ms, faster than 35.67% of C++ online submissions for Pascal's Triangle.
// Memory Usage: 7.4 MB, less than 33.07% of C++ online submissions for Pascal's Triangle.
class Solution {
 public:
  vector<vector<int>> generate(int num_rows) {
    // base case
    if (num_rows == 1) return vector<vector<int>>({{1}});

    // build up to the current level
    vector<vector<int>> res = generate(num_rows - 1);
    vector<int> last_row;

    // generate the n elements of the nth row
    for (int i = 0; i < num_rows; ++i) {
      int l = (i >= 1) ? res[num_rows - 2][i - 1]: 0;
      int r = (i < (num_rows - 1)) ? res[num_rows - 2][i]: 0;
      last_row.push_back(l + r);
    }

    res.push_back(last_row);
    return res;
  }
};

File: leetcode/118.py
"""
Runtime: 24 ms, faster than 94.44% of Python3 online submissions for Pascal's Triangle.
Memory Usage: 14.2 MB, less than 80.67% of Python3 online submissions for Pascal's Triangle.
"""


class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res = [[1]]
        if numRows == 1:
            return res

        for _ in range(numRows - 1):
            tmp = []
            curr = [0] + res[-1] + [0]
            for i in range(1, len(curr)):
                a, b = curr[i], curr[i - 1]
                tmp += [a + b]

            res.append(tmp)

        return res
